### ISSUE TITLE
Too many recalculations of route

### ISSUE THREAD

**Issue by Spielmops** [osmandapp/OsmAnd#16315]
Version 4.2.7:
Previously, when I couldn't follow a route further and had to deviate, I'd got a short silence, then "You're deviating from route," then silence, then "Route recalculated, distance, etc." in the driving direction. The app often tried to guide me back, but without an announcement.

Now with version 4.3.5, the app announces this after 120 meters: "Route recalculated, distance..." etc. always one after the other, often so fast that the announcement is no longer complete. And again and again... Totally annoying. And the navigation always just wants to go back to the old route. And this continues until the app finds a new route which is shorter than the previous one.

My wish: an announcement "You have deviated from the route for ... meters", then the route will be recalculated as quickly as possible and in the direction of travel!

Spielmops

**Comment by Zirochkabila** [comment]
@Spielmops 
Please clarify:
1. The device you use, OS version
2. Route, start and finish coordinates 
3. The presence of traffic restrictions
4. A short video and screenshots where we could see the error
5. Logcat file: Menu > Help > Send logcat log

**Comment by Spielmops** [comment]
1. Cubot Max, LineageOS 14, Samsung S9+, LineageOS 18, Samsung Tab S6 Lite, LineageOS 19
2. independant of rout start and stop
3. no restrictions
4. I'm driving, so no video
5. Logcat of which device?

If I tell you something about a difference, why do you need any info about device, OS, route start or stop and so on? 

The difference is: it has been OK, it is with the new Osmand-Version not OK any more.

If you need more, then forget it and close this issue ...

Spielmops

**Comment by vshcherb** [comment]
Recorded GPX track is needed at least part of it and a destination point, so we could simulate navigation part and  test it.

**Comment by vshcherb** [comment]
> If you need more, then forget it and close this issue ...

The issue could be fixed only if it's reproducible and navigation issues are mostly location specific, so if you would like to get it fixed we need more details. If you want us to be notified of possible bug, then you can close the issue.

**Comment by Spielmops** [comment]
https://hartmut.homelinux.org/Downloads/2023-01-21_11-41_Sat.gpx
gpx is not accepted here ...

The destination point is, where the track ends. The app wanted me to use the motorway A27 up to Walsroder Dreieck and then A7 back to Dorfmark.

Spielmops

That is not the only track ...

**Comment by sonora** [comment]
The issue description sounds a bit like my observation here https://github.com/osmandapp/OsmAnd/issues/14466#issuecomment-1356832126.

**Comment by vshcherb** [comment]
Then we definitely need to reproduce it shouldn't be complicated with recorded continuous gpx

**Comment by DmitryAlexei** [comment]
Tried to reproduce this issue with provided gpx track. The main problem is that you cannot deviate from the route in the simulation.

As for me, I get a lot of messages about recalculating the route when I drive along the "fuel-efficient" route instead of "short on time" (that is set by navigation settings). The difference between them is about 10 km and 7-8 min. So messages about “recalculation” I can hear about 3-4 min.

Announce route recalculation option can also be disabled in Voice prompts section of Navigation settings

**Comment by vshcherb** [comment]
@DmitryAlexei well we can create an artificial track to simulate deviations but indeed it's hard to understand is it on purpose or not. 
As I remember we silence after 3rd attempt of recalculation for a minute or so... However I don't see that in the code only recent change didn't change anything (https://github.com/osmandapp/OsmAnd/commit/3f3c288083ef75a931abb6447e7c75f24f683093). So I suspect the issue is not in voice commands but in route recalculation itself - which is also configurable sensitivity and probably some default parameters has changed?



https://github.com/osmandapp/OsmAnd/blame/master/OsmAnd/src/net/osmand/plus/routing/VoiceRouter.java#L23

@sonora waitAnnouncedOffRoute *= 2.5; - keeps multiplying forever ? so for an application start you would only hear 3-4 times. I guess it should be correlated with back on route and should reset somehow...

**Comment by DmitryAlexei** [comment]
OsmAnd~ 4.4.0#14178mqta, released: 2023-02-15
Reproduced with simulation speed set to 162 mi/h (260 km/h) and only in the middle of the track.
Minimal distance to recalculate route set to 120 m (default)
Also enabled recalculation of route _In case of reverse direction_ (default) and _Smart route recalculation_ (default)


https://user-images.githubusercontent.com/116079541/219011191-7a9b1344-cbaf-4561-be71-24ae38264ee6.mp4


https://user-images.githubusercontent.com/116079541/219012048-444295aa-b84c-4e34-b9e5-a3e24b6b4320.mp4

**Comment by DmitryAlexei** [comment]
It seems like numerous messages about route recalculation reproduce in a situation of deviation from the route when:
- you are moving in the opposite direction of your route
- there are no roads around suitable for detour in direction of movement.

**Comment by Spielmops** [comment]
As a reminder from top of this thread:

> Version 4.2.7:
> Previously, when I couldn't follow a route further and had to deviate, I'd got a short silence, then "You're deviating from route," then silence, then "Route recalculated, distance, etc." in the driving direction. The app often tried to guide me back, but without any announcement.

And: I use the app very often, because I travel with a motorhome through Europe. So I can tell: something has changed and not from my side. My navigation settings are functional as can be for me, so I don't have to change anything. I just changed the app-version ...
What has changed in the app?

Spielmops

**Comment by sonora** [comment]
> waitAnnouncedOffRoute *= 2.5; - keeps multiplying forever ?

@vshcherb Indeed! Looks like this went unnoticed by both of us through many refinements, and since the very creation of that method in 2013 (https://github.com/osmandapp/OsmAnd/commit/ad528473b12c32ca78a079abe2218673433f78bc). Shame on us, but we were both young! 😉 

I think the correct condition for a reset is that _the user follows a route again **in forward direction**_. In that sense it is connected to this user story https://github.com/osmandapp/OsmAnd/issues/16450#issuecomment-1428530941. Will investigate, but we may first want to wait if we implement that user story or what the proper fix for #16450 will be, so we can set the correct rest condition here.

**Comment by sonora** [comment]
PS: I see no changes in those sections of _Voice Router_, _RoutingHelper_, or _RouteRecalculationHelper_ which govern when we announce or silence announcements due to deviations from the route or moving in the wrong direction.

I suspect we may perhaps have changed some parameterization around our route recalculation triggers, but cannot currently locate where.

In any case I suspect that #11206, #14466, #16450, and this one may all have the same root cause.

**Comment by vshcherb** [comment]
Probably merged with https://github.com/osmandapp/OsmAnd/pull/16548

**Comment by Zirochkabila** [comment]
OsmAnd~ 4.4.0#14217mqta, released: 2023-03-04
04.03 master 
Done
